src/qibolab/instruments/zhinst/executor.py

```diff
@@ -0,0 +1,781 @@
+                zhs.values /= max(zhs.values)
+                play_parameters["amplitude"] = zhs
+            if p is Parameter.duration:
+                play_parameters["length"] = zhs
+            if p is Parameter.relative_phase:
+                play_parameters["phase"] = zhs
+        if "phase" not in play_parameters:
+            play_parameters["phase"] = pulse.pulse.relative_phase
+
+        exp.play(signal=channel_name, pulse=pulse.zhpulse, **play_parameters)
+
+    @staticmethod
+    def rearrange_rt_sweepers(
+        sweepers: list[Sweeper],
+    ) -> tuple[np.ndarray, list[Sweeper]]:
+        """Rearranges sweepers from qibocal based on device hardware
+        limitations.
+
+        The only known limitation currently is that frequency sweepers must be applied before (on the outer loop) other
+        (e.g. amplitude) sweepers. Consequently, the only thing done here is to swap the frequency sweeper with the
+        first sweeper in the list.
+
+        Args:
+            sweepers (list[Sweeper]): Sweepers to rearrange.
+
+        Returns:
+            swapped_axis_pair (np.ndarray): array of shape (2,) containing the indices of the two swapped axes.
+            sweepers (list[Sweeper]): rearranged list of sweepers.
+        """
+        swapped_axis_pair = np.zeros(2, dtype=int)
+        sweepers_copy = sweepers.copy()
+        freq_sweeper = next(
+            iter(s for s in sweepers_copy if s.parameter is Parameter.frequency), None
+        )
+        if freq_sweeper:
+            freq_sweeper_idx = sweepers_copy.index(freq_sweeper)
+            sweepers_copy[freq_sweeper_idx] = sweepers_copy[0]
+            sweepers_copy[0] = freq_sweeper
+            swapped_axis_pair = np.array([0, freq_sweeper_idx])
+            log.warning("Sweepers were reordered")
+        return swapped_axis_pair, sweepers_copy
+
+    def sweep(self, qubits, couplers, sequence: PulseSequence, options, *sweepers):
+        """Play pulse and sweepers sequence."""
+
+        self.signal_map = {}
+        self.processed_sweeps = ProcessedSweeps(sweepers, qubits)
+        self.nt_sweeps, self.rt_sweeps = classify_sweepers(sweepers)
+        swapped_axis_pair, self.rt_sweeps = self.rearrange_rt_sweepers(self.rt_sweeps)
+        swapped_axis_pair += len(self.nt_sweeps)
+        # if using singleshot, the first axis contains shots,
+        # i.e.: (nshots, sweeper_1, sweeper_2)
+        # if using integration: (sweeper_1, sweeper_2)
+        if options.averaging_mode is AveragingMode.SINGLESHOT:
+            swapped_axis_pair += 1
+
+        self.frequency_from_pulses(qubits, sequence)
+
+        for sweeper in sweepers:
+            if sweeper.parameter in {Parameter.frequency, Parameter.amplitude}:
+                for pulse in sweeper.pulses:
+                    if pulse.type is PulseType.READOUT:
+                        self.acquisition_type = lo.AcquisitionType.SPECTROSCOPY
+
+        self.experiment_flow(qubits, couplers, sequence, options)
```

Comment:
Currently, the call stack is:
```py
sweep()
experiment_flow()
create_exp()
  sweep_recursion_nt()
    sweep_recursion_nt()
      ...
define_exp()
  sweep_recursion()
    sweep_recursion()
      ...
select_exp()
```
the one below nested in those above.

I like the modularity, but nesting too much is also problematic, to reconstruct the flow.
I'd try to make an effort to denest a bit.

And this could happen:
- after merging `sweep()` with `play()`, you could in principle inline `experiment_flow()` with reduced duplication
- or keep `experiment_flow()`, but lift just the call to `create_exp()` to `sweep()`/`play()`
- if `sweep_recursion_nt()` and `sweep_recursion()` will be joined, you do not need to split the recursion in two pieces, just sort the sweepers to be consumed before starting the recursion